File: src/Passes.hs
module Passes (
  functionBodies,
  passBase,
  smbSwitchPass
) where

import AddressSpace (AddressSpace)
import qualified Assembly as I
import Data.Functor ((<&>))
import Data.Map (Map, (!))
import qualified Data.Map as Map
import Data.Word (Word16)

data Parser = Parser {
  postProcess :: [Word16] -> Map Word16 (Map Word16 I.Instruction) -> Map Word16 (Map Word16 I.Instruction),
  readInstruction :: Word16 -> AddressSpace -> I.Instruction
}

type Pass = Parser -> Parser

functionBodies :: Parser -> [Word16] -> AddressSpace -> Map Word16 (Map Word16 I.Instruction)
functionBodies p offs addr = postProcess p offs funcs
  where funcs = I.functionBodiesWithParser (\off -> readInstruction p off addr) offs

passBase :: Parser
passBase = Parser {
  postProcess = const id,
  readInstruction = I.readInstruction
}

smbSwitchPass :: Pass
smbSwitchPass underlying = Parser {
    postProcess = postProcess_,
    readInstruction = readInstruction_
  }
  where readInstruction_ off = do
          inst <- readInstruction underlying off
          case inst of
            (I.Absolute _ I.JSR 0x8e04) -> do
              addrs <- mapM I.readAddress [off + 3, (off + 5)..] <&> takeWhile (>= 0x8000)
              return $ I.Switch off I.SWA 3 addrs
            i -> return i
        postProcess_ offs funcs = postProcess underlying offs $ I.functionBodiesWithParser parse offs
          where under = postProcess underlying offs funcs
                filterAddrs (I.Switch off I.SWA len addrs) = I.Switch off I.SWA len $ map snd $ takeWhile (\(i, _) -> isNotCode $ 2 * i + off + 3) $ zip [0..] addrs
                filterAddrs i = i
                isNotCode a = null $ instructionsMatching (\i -> a < I.nextAddr i && a + 1 >= I.offset i) under
                repaired = mapInstructions filterAddrs under
                flattened = foldr Map.union Map.empty $ Map.elems repaired
                parse = (!) flattened

mapInstructions :: (I.Instruction -> I.Instruction) -> Map Word16 (Map Word16 I.Instruction) -> Map Word16 (Map Word16 I.Instruction)
mapInstructions f = Map.map (Map.map f)

instructionsMatching :: (I.Instruction -> Bool) -> Map Word16 (Map Word16 I.Instruction) -> [I.Instruction]
instructionsMatching f = filter f . instructions

instructions :: Map Word16 (Map Word16 I.Instruction) -> [I.Instruction]
instructions funcs = do
  body <- Map.elems funcs
  Map.elems body

File: parser/src/passes.rs
use std::mem::swap;

use crate::{
    assembly::{
        function_bodies_with_parser, irq_address, nmi_address, reset_address, Functions,
        Instruction, Instructions, Opcode,
    },
    mapper::Mapper,
};

pub trait Parser {
    fn read_instruction(&self, buf: &dyn Mapper, addr: u16) -> Instruction;
    fn post_process(&self, roots: &[u16], functions: &mut Functions);

    fn parse_roots(&self, buf: &dyn Mapper, roots: &[u16]) -> Functions {
        let mut funcs = function_bodies_with_parser(|addr| self.read_instruction(buf, addr), roots);
        self.post_process(roots, &mut funcs);
        funcs
    }

    fn parse(&self, buf: &dyn Mapper) -> Functions {
        self.parse_roots(
            buf,
            &[reset_address(buf), nmi_address(buf), irq_address(buf)],
        )
    }
}

pub struct BaseParser;

impl Parser for BaseParser {
    fn read_instruction(&self, buf: &dyn Mapper, addr: u16) -> Instruction {
        Instruction::read(buf, addr)
    }

    fn post_process(&self, _roots: &[u16], _functions: &mut Functions) {}
}

pub struct SelfLoopPass {
    inner: Box<dyn Parser>,
}

impl SelfLoopPass {
    pub fn new(inner: Box<dyn Parser>) -> Self {
        Self { inner }
    }

    pub fn with_inner(inner: impl Parser + 'static) -> Self {
        Self::new(Box::new(inner))
    }
}

impl Parser for SelfLoopPass {
    fn read_instruction(&self, buf: &dyn Mapper, addr: u16) -> Instruction {
        match self.inner.read_instruction(buf, addr) {
            Instruction::Absolute {
                loc,
                opcode: Opcode::JMP,
                addr,
            } if loc == addr => Instruction::Implied {
                loc,
                opcode: Opcode::SLP,
            },
            i => i,
        }
    }

    fn post_process(&self, roots: &[u16], functions: &mut Functions) {
        self.inner.post_process(roots, functions);
    }
}

pub struct SmbSwitchPass {
    inner: Box<dyn Parser>,
}

impl SmbSwitchPass {
    pub fn new(inner: Box<dyn Parser>) -> Self {
        Self { inner }
    }

    pub fn with_inner(inner: impl Parser + 'static) -> Self {
        Self::new(Box::new(inner))
    }
}

impl Parser for SmbSwitchPass {
    fn read_instruction(&self, buf: &dyn Mapper, addr: u16) -> Instruction {
        match self.inner.read_instruction(buf, addr) {
            Instruction::Absolute {
                opcode: Opcode::JSR,
                addr: 0x8e04,
                ..
            } => {
                let mut targets = vec![];
                for i in 0.. {
                    let target = buf.read_address(addr + 3 + (2 * i));
                    if target < 0x8000 {
                        break;
                    }
                    targets.push(target);
                }
                Instruction::Switch {
                    loc: addr,
                    opcode: Opcode::SWA,
                    targets,
                }
            }
            i => i,
        }
    }

    fn post_process(&self, roots: &[u16], functions: &mut Functions) {
        self.inner.post_process(roots, functions);
        let flattened: Instructions = functions
            .iter()
            .flat_map(|(_, insts)| insts.values())
            .map(|inst| {
                if let Instruction::Switch {
                    loc,
                    opcode,
                    targets,
                    ..
                } = inst
                {
                    let new_targets = targets
                        .iter()
                        .cloned()
                        .take_while(|i| is_code(functions, *i))
                        .collect();
                    (
                        *loc,
                        Instruction::Switch {
                            loc: *loc,
                            opcode: *opcode,
                            targets: new_targets,
                        },
                    )
                } else {
                    (inst.offset(), inst.clone())
                }
            })
            .collect();
        let mut trimmed_funcs = function_bodies_with_parser(
            |addr| {
                flattened
                    .get(&addr)
                    .cloned()
                    .unwrap_or(Instruction::Unknown {
                        loc: addr,
                        opcode: 0,
                    })
            },
            roots,
        );
        self.inner.post_process(roots, &mut trimmed_funcs);
        swap(functions, &mut trimmed_funcs);
    }
}

fn is_code(funcs: &Functions, addr: u16) -> bool {
    funcs
        .values()
        .filter(|insts| {
            insts
                .range(0..=addr)
                .next_back()
                .filter(|(_, floor)| addr < floor.next_addr())
                .is_some()
        })
        .next()
        .is_some()
}
